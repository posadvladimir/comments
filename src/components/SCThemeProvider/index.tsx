
import { ThemeProvider } from 'styled-components';

import type { TChildrenProps } from 'src/ts/types/children';

import theme from 'src/styles/theme';

const SCThemeProvider: React.FC<TChildrenProps> = ({ children }) => (
    <ThemeProvider theme={theme}>
        {children}
    </ThemeProvider>
);

export default SCThemeProvider;