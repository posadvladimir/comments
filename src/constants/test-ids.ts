
const LOADER_TEST_ID = 'loader-test-id';
const BUTTON_TEST_ID = 'button-test-id';
const LIKE_COUNTER_TEST_ID = 'like-counter-test-id';
const LIKE_COUNTER_BUTTON_TEST_ID = 'like-counter-button-test-id';
const LIKE_COUNTER_TOTAL_TEST_ID = 'like-counter-total-test-id';
const COMMENT_TEST_ID = 'comment-test-id';
const ERROR_TEST_ID = 'error-test-id';

export {
    LOADER_TEST_ID,
    BUTTON_TEST_ID,
    LIKE_COUNTER_TEST_ID,
    LIKE_COUNTER_BUTTON_TEST_ID,
    COMMENT_TEST_ID,
    LIKE_COUNTER_TOTAL_TEST_ID,
    ERROR_TEST_ID,
};