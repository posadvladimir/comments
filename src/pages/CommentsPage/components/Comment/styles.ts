
import { styled } from 'styled-components';

const AVATAR_SIZE = 68;
const AVATAR_SIZE_PHONE = 40;
const AVATAR_MARGIN = 20;

export const CommentWrapper = styled.div<{ $addMarginLeft: boolean }>`
    margin-top: 32px;
    ${({ $addMarginLeft }) => $addMarginLeft ? `margin-left: 34px;` : ''}

    @media ${props => props.theme.media.phone} {
        margin-top: 24px;
        ${({ $addMarginLeft }) => $addMarginLeft ? `margin-left: 20px;` : ''}
    }
`;

export const CommentHeader = styled.div`
    display: flex;
    align-items: center;
    width: 100%;
`;

export const AuthorAvatar = styled.img`
    object-fit: cover;
    width: ${AVATAR_SIZE}px;
    height: ${AVATAR_SIZE}px;
    border-radius: 50%;
    margin-right: ${AVATAR_MARGIN}px;

    @media ${props => props.theme.media.phone} {
        width: ${AVATAR_SIZE_PHONE}px;
        height: ${AVATAR_SIZE_PHONE}px;
    }
`;

export const CommentDetails = styled.div`
    display: flex;
    flex-direction: column;
    flex-grow: 1;
    margin: 12.5px 0;

    @media ${props => props.theme.media.phone} {
        margin: 0;
    }
`;

export const CommentCreationTime = styled.span`
    color: #8297AB;
`;

export const CommentText = styled.p`
    margin-left: ${AVATAR_SIZE + AVATAR_MARGIN}px;
    word-wrap: break-word;

    @media ${props => props.theme.media.phone} {
        margin-top: 8px;
        margin-left: ${AVATAR_SIZE_PHONE + AVATAR_MARGIN}px;
    }
`;