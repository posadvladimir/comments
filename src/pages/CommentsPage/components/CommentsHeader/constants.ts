
import type { TDeclensionArray } from '../../common/ts/types/declensions';

export const COMMENT_DECLENSIONS: TDeclensionArray = [
    'комментарий',
    'комментария',
    'комментариев'
];