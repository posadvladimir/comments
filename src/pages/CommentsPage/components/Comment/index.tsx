
import type { TCommentDataForRender } from '../../common/ts/types/comments';

import { COMMENT_TEST_ID } from 'src/constants/test-ids';

import { BoldSpan } from '../../common/styles/boldSpan';
import * as S from './styles';

import LikeCounter from '../LikeCounter';

const Comment: React.FC<TCommentDataForRender> = ({
    authorName,
    authorAvatar,
    likes,
    text,
    creationTime,
    parent,
}) => (
    <S.CommentWrapper
        $addMarginLeft={Boolean(parent)}
        data-testid={COMMENT_TEST_ID}    
    >
        <S.CommentHeader>
            <S.AuthorAvatar
                src={authorAvatar}
                alt="author avatar"
            />
            <S.CommentDetails>
                <BoldSpan>{authorName}</BoldSpan>
                <S.CommentCreationTime>{creationTime}</S.CommentCreationTime>
            </S.CommentDetails>
            <LikeCounter likes={likes} />
        </S.CommentHeader>
        <S.CommentText>{text}</S.CommentText>
    </S.CommentWrapper>
);

export default Comment;