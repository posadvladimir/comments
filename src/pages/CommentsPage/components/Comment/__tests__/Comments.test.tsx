
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';

import { LIKE_COUNTER_TEST_ID } from 'src/constants/test-ids';

import { runSnapshotTest } from 'src/common/tests/snapshot';

import SCThemeProvider from 'src/components/SCThemeProvider';
import Comment from '..';

const mockCommentData = {
    authorName: 'Han Solo',
    authorAvatar: 'han-solo.jpg',
    likes: 10,
    text: 'Piu piu',
    creationTime: '3 часа назад',
    parent: null,
};

const getCommentComponent = (parent: number | null) => (
    <SCThemeProvider>
        <Comment
            {...mockCommentData}
            parent={parent}
        />
    </SCThemeProvider>
);

const renderComment = (parent: number | null):void => {
    render(getCommentComponent(parent));
};

describe('Компонент Comment', () => {
    test('отображается имя автора', () => {
        renderComment(null);

        const authorNameElement = screen.getByText(mockCommentData.authorName);

        expect(authorNameElement).toBeInTheDocument();
    });

    test('отображается текст комментария', () => {
        renderComment(null);

        const commentTextElement = screen.getByText(mockCommentData.text);

        expect(commentTextElement).toBeInTheDocument();
    });

    test('отображается время создания комментария', () => {
        renderComment(null);

        const creationTimeElement = screen.getByText(mockCommentData.creationTime);

        expect(creationTimeElement).toBeInTheDocument();
    });

    test('отображается компонент LikeCounter', () => {
        renderComment(null);

        const likeCounterElement = screen.getByTestId(LIKE_COUNTER_TEST_ID);

        expect(likeCounterElement).toBeInTheDocument();
    });

    test('не добавляется margin-left, если значение prop parent falsy', () => {
        const tree = renderer.create(
            <SCThemeProvider>
                <Comment {...mockCommentData} />
            </SCThemeProvider>
        ).toJSON();

        expect(tree).not.toHaveStyleRule('margin-left', '34px');
    });

    test('добавляется margin-left равное 34px, если значение prop parent определенно', () => {
        const tree = renderer.create(
            <SCThemeProvider>
                <Comment
                    {...mockCommentData}
                    parent={200}
                />
            </SCThemeProvider>
        ).toJSON();

        expect(tree).toHaveStyleRule('margin-left', '34px');
    });

    runSnapshotTest(getCommentComponent(null));

    // отображается комментарий с левым отступом (дочерний)
    runSnapshotTest(getCommentComponent(200));
});