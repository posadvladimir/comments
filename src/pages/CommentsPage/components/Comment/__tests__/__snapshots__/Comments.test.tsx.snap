// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Компонент Comment соотвествует снепшоту 1`] = `
.c4 {
  line-height: 22px;
  font-weight: 700;
}

.c0 {
  margin-top: 32px;
}

.c1 {
  display: flex;
  align-items: center;
  width: 100%;
}

.c2 {
  object-fit: cover;
  width: 68px;
  height: 68px;
  border-radius: 50%;
  margin-right: 20px;
}

.c3 {
  display: flex;
  flex-direction: column;
  flex-grow: 1;
  margin: 12.5px 0;
}

.c5 {
  color: #8297AB;
}

.c10 {
  margin-left: 88px;
  word-wrap: break-word;
}

.c6 {
  display: flex;
  font-size: 15px;
  font-weight: 700;
  line-height: 150%;
}

.c7 {
  cursor: pointer;
}

.c7[disabled] {
  cursor: auto;
}

.c8 {
  width: 22px;
  height: 22px;
}

.c9 {
  margin-left: 8px;
}

@media (max-width: 430px) {
  .c0 {
    margin-top: 24px;
  }
}

@media (max-width: 430px) {
  .c2 {
    width: 40px;
    height: 40px;
  }
}

@media (max-width: 430px) {
  .c3 {
    margin: 0;
  }
}

@media (max-width: 430px) {
  .c10 {
    margin-top: 8px;
    margin-left: 60px;
  }
}

@media (max-width: 430px) {
  .c6 {
    font-size: inherit;
  }
}

<div
  className="c0"
  data-testid="comment-test-id"
>
  <div
    className="c1"
  >
    <img
      alt="author avatar"
      className="c2"
      src="han-solo.jpg"
    />
    <div
      className="c3"
    >
      <span
        className="c4"
      >
        Han Solo
      </span>
      <span
        className="c5"
      >
        3 часа назад
      </span>
    </div>
    <div
      className="c6"
      data-testid="like-counter-test-id"
    >
      <button
        className="c7"
        data-testid="like-counter-button-test-id"
        onClick={[Function]}
      >
        <img
          alt="like icon"
          className="c8"
          src="red-stroke.png"
        />
      </button>
      <span
        className="c9"
        data-testid="like-counter-total-test-id"
      >
        10
      </span>
    </div>
  </div>
  <p
    className="c10"
  >
    Piu piu
  </p>
</div>
`;

exports[`Компонент Comment соотвествует снепшоту 2`] = `
.c4 {
  line-height: 22px;
  font-weight: 700;
}

.c0 {
  margin-top: 32px;
  margin-left: 34px;
}

.c1 {
  display: flex;
  align-items: center;
  width: 100%;
}

.c2 {
  object-fit: cover;
  width: 68px;
  height: 68px;
  border-radius: 50%;
  margin-right: 20px;
}

.c3 {
  display: flex;
  flex-direction: column;
  flex-grow: 1;
  margin: 12.5px 0;
}

.c5 {
  color: #8297AB;
}

.c10 {
  margin-left: 88px;
  word-wrap: break-word;
}

.c6 {
  display: flex;
  font-size: 15px;
  font-weight: 700;
  line-height: 150%;
}

.c7 {
  cursor: pointer;
}

.c7[disabled] {
  cursor: auto;
}

.c8 {
  width: 22px;
  height: 22px;
}

.c9 {
  margin-left: 8px;
}

@media (max-width: 430px) {
  .c0 {
    margin-top: 24px;
    margin-left: 20px;
  }
}

@media (max-width: 430px) {
  .c2 {
    width: 40px;
    height: 40px;
  }
}

@media (max-width: 430px) {
  .c3 {
    margin: 0;
  }
}

@media (max-width: 430px) {
  .c10 {
    margin-top: 8px;
    margin-left: 60px;
  }
}

@media (max-width: 430px) {
  .c6 {
    font-size: inherit;
  }
}

<div
  className="c0"
  data-testid="comment-test-id"
>
  <div
    className="c1"
  >
    <img
      alt="author avatar"
      className="c2"
      src="han-solo.jpg"
    />
    <div
      className="c3"
    >
      <span
        className="c4"
      >
        Han Solo
      </span>
      <span
        className="c5"
      >
        3 часа назад
      </span>
    </div>
    <div
      className="c6"
      data-testid="like-counter-test-id"
    >
      <button
        className="c7"
        data-testid="like-counter-button-test-id"
        onClick={[Function]}
      >
        <img
          alt="like icon"
          className="c8"
          src="red-stroke.png"
        />
      </button>
      <span
        className="c9"
        data-testid="like-counter-total-test-id"
      >
        10
      </span>
    </div>
  </div>
  <p
    className="c10"
  >
    Piu piu
  </p>
</div>
`;
